cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc

```diff
@@ -1,17 +1,165 @@
 #include "tiledb_cxx_array_util.h"
-//#include "tiledb_cxx_json.hpp"
-//#include "tiledb_cxx_string_util.h"
+
 
 namespace tiledb {
 
-	std::string ArrayUtil::get_tiledb_version()
-	{
-		std::vector<int> t = tiledb::version();
-		return std::to_string(t[0]) + "." + std::to_string(t[1]) + "." + std::to_string(t[2]);
-	}//std::string ArrayUtil::get_tiledb_version()
+std::string ArrayUtil::get_tiledb_version()
+{
+  std::vector<int> t = tiledb::version();
+  return std::to_string(t[0]) + "." + std::to_string(t[1]) + "." + std::to_string(t[2]);
+}//std::string ArrayUtil::get_tiledb_version()
+
+int ArrayUtil::export_file_to_path(const std::string& file_uri, const std::string& output_path, uint64_t buffer_size, const std::shared_ptr<tiledb::Context>& ctx)
+{
+  const std::string METADATA_SIZE_KEY = "file_size";
+  const std::string FILE_ATTRIBUTE_NAME = "contents";
+  const std::string METADATA_ORIGINAL_FILE_NAME = "original_file_name";
+  try {
+
+    std::shared_ptr<Array> array = std::shared_ptr<Array>(new Array(ctx, file_uri, tiledb::QueryType::TILEDB_READ));
+
+    //get file_size from metadata
+    uint32_t v_num;
+    const void* v_r;
+    tiledb::DataType file_size_datatype = tiledb::DataType::TILEDB_UINT64;
+    array->get_metadata(METADATA_SIZE_KEY, &file_size_datatype, &v_num, &v_r);
+    uint64_t file_size = *((const int64_t*)v_r);
+    if(buffer_size==0) {
+      buffer_size = file_size;
+    }
+    if (buffer_size <= 0) {
+      std::cout << "ArrayUtil::export_file_to_path, wrong buffer_size!" << std::endl;
+      return -1;
+    }
+
+    std::string output_file = output_path;
+    if(output_file.empty()) {
+      tiledb::DataType file_name_datatype = tiledb::DataType::TILEDB_STRING_ASCII;
+      uint32_t name_size = 0;
+      const char* original_name;
+      array->get_metadata(METADATA_ORIGINAL_FILE_NAME,&file_name_datatype,&name_size,reinterpret_cast<const void**>(&original_name));
+      output_file = std::string(original_name, name_size);
+    }
+    
+    std::shared_ptr<tiledb::VFS> vfs = std::shared_ptr<tiledb::VFS>(new tiledb::VFS(ctx));
+    tiledb::VFS::filebuf filebuf = tiledb::VFS::filebuf(vfs);
+    filebuf.open(output_file, std::ios::out);
+    std::ostream os(&filebuf);   
+    
+    uint64_t offset = 0;
+    std::array<uint64_t, 2> subarray = { offset, offset + buffer_size - 1 };
+
+    Query query(ctx, array, tiledb::QueryType::TILEDB_READ);
+    query.set_layout(tiledb::LayoutType::TILEDB_ROW_MAJOR)
+      .set_subarray(subarray);
+    std::vector<uint8_t> data_buffer;
```

Comment:
This can be initialized with `buffer_size` as `data_buffer(buffer_size)`.

(In the future we should use a dedicated buffer class which does not initialize write zeros to the elements, because there's not need and it is expensive)

Reply:
changed to data_buffer(buffer_size)

Reply:
thanks